### PR TITLE
Configuration Overhaul

### PR DESCRIPTION
**Pull request by ejfitzgerald**
The main focus of this PR is to clean up the `main.cpp` of Constellation which has grown to be a bit out of control.

I have introduced a new `Setting` class which is a new abstraction of a command line parameter and a potential environment variable. In the future, this can also be mapped to a simple configuration structure.

### CODE REVIEW

File: apps/constellation/key_generator.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  bool      certificate_loaded = false;
+
+  // Step 1. Attempt to load the existing key
+  {
+    std::ifstream input_file(key_path, std::ios::in | std::ios::binary);
+
+    if (input_file.is_open())
+    {
+      fetch::byte_array::ByteArray private_key_data;
+      private_key_data.Resize(ECDSASigner::PrivateKey::ecdsa_curve_type::privateKeySize);
```
Comment (wkaluza) [inline apps/constellation/key_generator.cpp:71]: I think ByteArray has a ctor which takes a size_t argument and calls resize for you

File: apps/constellation/key_generator.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+namespace fetch {
+
+using fetch::crypto::Prover;
+using fetch::crypto::ECDSASigner;
+
+using SignerPtr = std::shared_ptr<ECDSASigner>;
+using ProverPtr = std::shared_ptr<Prover>;
+
+char const *LOGGING_NAME         = "KeyGenerator";
+char const *DEFAULT_KEY_FILENAME = "p2p.key";
```
Comment (wkaluza) [inline apps/constellation/key_generator.cpp:33]: Probably meaningless, but these may be constexpr

File: apps/constellation/manifest_builder.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+ * Create the manifest from an environment variable based configuration
+ *
+ * @param manifest The manifest to be populated
+ * @return true if successful, otherwise false
+ */
+bool LoadManifestFromEnvironment(Manifest &manifest)
+{
+  bool success{false};
+
+  char const *manifest_data = std::getenv("CONSTELLATION_MANIFEST");
```
Comment (wkaluza) [inline apps/constellation/manifest_builder.cpp:102]: rather than calling std::getenv directly, I'd like to see a thin wrapper around it which we own and can mock out. Aids testability. Rather than a string, the wrapper could take an enum class argument like ConstellationEnvVar, and would look up the right env var string name based on the arg value. This would also have the benefit of putting all env vars constellation depends on in one place - we only have a couple for now, but it's easy to imagine their number growing over time, and the code getting peppered with magic strings like this one.

File: apps/constellation/settings.hpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+  /// @name Proof of Stake
+  /// @{
+  settings::Setting<bool>        proof_of_stake;
+  settings::Setting<std::string> beacon_address;
+  /// @}
+
+  // Operators
+  Settings &operator=(Settings const &) = delete;
+  Settings &operator=(Settings &&) = delete;
```
Comment (wkaluza) [inline apps/constellation/settings.hpp:112]: Hissss....

File: apps/constellation/settings.hpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+#include <cstdint>
+#include <string>
+#include <vector>
+
+namespace fetch {
+
+/**
+ * Command line / Environment variable settings
+ */
+class Settings : protected settings::SettingCollection
```
Comment (wkaluza) [inline apps/constellation/settings.hpp]: inheritance may be private?

File: libs/settings/include/settings/setting_collection.hpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+class SettingBase;
+
+/**
+ * A very simple collection of pointers to settings object. The use must make sure that the
+ * lifetime of the settings is longer than this container
+ */
+class SettingCollection
+{
+public:
+  using Settings = std::vector<SettingBase *>;
```
Comment (wkaluza) [inline libs/settings/include/settings/setting_collection.hpp:36]: may this be a vector of unique_ptrs? Lifetime concerns would go away if so.

File: libs/settings/include/settings/detail/setting_list.hpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
+#pragma once
```
Comment (wkaluza) [inline libs/settings/include/settings/detail/setting_list.hpp]: this file is usable more generally; namespacing it as a detail of settings may need a rethink. Could it find a home together with the EndsWith string helper that you recently wrote?

Unrelatedly, setting_list.hpp as a name may be confusing, especially given that setting_collection.hpp exists and contains very different functionality.

File: libs/settings/tests/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+#
+# F E T C H   M I N E R   T E S T S
```
Comment (wkaluza) [inline libs/settings/tests/CMakeLists.txt]: Copypasta error.

File: libs/settings/tests/unit/setting_collection_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  {
+    return argv_.data();
+  }
+
+private:
+  std::vector<std::string> args_;
+  std::vector<char *>      argv_;
+};
+
+TEST(SettingCollectionTests, SimpleCheck)
```
Comment (wkaluza) [inline libs/settings/tests/unit/setting_collection_tests.cpp]: This is nice, but while we're here, can we test stuff typical edge cases? Passing in no settings, passing in a misspelled option, passing in `xyz` for the number of lanes?

File: libs/settings/tests/unit/setting_collection_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  Setting<uint32_t>    lanes{collection, "lanes", 0, ""};
+  Setting<std::string> name{collection, "name", "default", ""};
+
+  SystemArgAdapter arguments({"-lanes", "256", "-name", "foo-bar-baz"});
+
+  collection.UpdateFromArgs(arguments.argc(), arguments.argv());
+
+  EXPECT_EQ(lanes.value(), 256);
+  EXPECT_EQ(name.value(), "foo-bar-baz");
+}
```
Comment (wkaluza) [inline libs/settings/tests/unit/setting_collection_tests.cpp:178]: I think we must test the interaction etween env variables and their corresponding config/cli options, and which take precedence when. The wrapper I suggested above may be a good first step.

**Review by wkaluza — CHANGES_REQUESTED** [review]
Looks good, some niggles and suggested tests.

File: apps/constellation/key_generator.cpp (hunk 1 above)
Comment (ejfitzgerald) [reply apps/constellation/key_generator.cpp:71]: Nope that is not the case (unfortunately)

File: apps/constellation/key_generator.cpp (hunk 2 above)
Comment (ejfitzgerald) [reply apps/constellation/key_generator.cpp:33]: `DEFAULT_KEY_FILENAME` certainly can't because we need to be able to make a pointer to it! 😁

File: apps/constellation/settings.hpp (hunk 4 above)
Comment (ejfitzgerald) [reply apps/constellation/settings.hpp:112]: ?

File: apps/constellation/settings.hpp (hunk 5 above)
Comment (ejfitzgerald) [reply apps/constellation/settings.hpp]: Yep it could be in this case. `protected` seems a good idea in case we wanted to have derived configs in the future.

File: libs/settings/include/settings/setting_collection.hpp (hunk 6 above)
Comment (ejfitzgerald) [reply libs/settings/include/settings/setting_collection.hpp:36]: Specifically the design doesn't want to control the lifetime of these objects. The `unique_ptr` would change this quite drastically. The closest version compatible version would be to use the `shared_ptr` or to use `std::reference_wrapper`. I did initially look at the reference wrapper and decided that it was more effort than was required.

The basic design here is that everything is designed to be very static so having this pointer vector is fine IMHO.

File: libs/settings/include/settings/detail/setting_list.hpp (hunk 7 above)
Comment (ejfitzgerald) [reply libs/settings/include/settings/detail/setting_list.hpp]: Yes it might be useful. I would suggest that we find that use case first and then generalise it. I am happy for filename suggestions but this seems fine to me

File: libs/settings/tests/CMakeLists.txt (hunk 8 above)
Comment (ejfitzgerald) [reply libs/settings/tests/CMakeLists.txt]: done

File: apps/constellation/settings.hpp (hunk 4 above)
Comment (wkaluza) [reply apps/constellation/settings.hpp:112]: What we talked about - assignment operators far away from constructors. :) I know that's the style, but I refuse to like it.

File: apps/constellation/settings.hpp (hunk 5 above)
Comment (wkaluza) [reply apps/constellation/settings.hpp]: Let's make it private for now, until something actually derives from it.

File: libs/settings/include/settings/setting_collection.hpp (hunk 6 above)
Comment (wkaluza) [reply libs/settings/include/settings/setting_collection.hpp:36]: ok

File: apps/constellation/key_generator.cpp (hunk 1 above)
Comment (wkaluza) [reply apps/constellation/key_generator.cpp]: Seems like it would be useful. Consider adding it as part of this change.

Comment (wkaluza) [reply apps/constellation/key_generator.cpp:71]: Seems like it would be useful. Consider adding it as part of this change?

File: libs/settings/include/settings/detail/setting_list.hpp (hunk 7 above)
Comment (wkaluza) [reply libs/settings/include/settings/detail/setting_list.hpp]: csv_string_helpers.hpp seems truthful enough

File: apps/constellation/key_generator.cpp (hunk 1 above)
Comment (ejfitzgerald) [reply apps/constellation/key_generator.cpp:71]: Not for this change. It would be a change in of itself since this pattern of create instance and resize is prevalent in the code base

File: libs/settings/tests/unit/setting_collection_tests.cpp (hunk 9 above)
Comment (ejfitzgerald) [reply libs/settings/tests/unit/setting_collection_tests.cpp]: Done

File: apps/constellation/manifest_builder.cpp (hunk 3 above)
Comment (ejfitzgerald) [reply apps/constellation/manifest_builder.cpp:102]: This is sort of a chicken and egg problem a little bit because it is hard to test the value of an environment variable. 

The enum approach would not work because the whole point of this setting change is to allow dynamically query based on the collection settings.

I will raise an issue to track this and I think that we can tackle this as another PR.

File: libs/settings/tests/unit/setting_collection_tests.cpp (hunk 10 above)
Comment (ejfitzgerald) [reply libs/settings/tests/unit/setting_collection_tests.cpp:178]: See comment above

File: apps/constellation/key_generator.cpp (hunk 1 above)
Comment (wkaluza) [reply apps/constellation/key_generator.cpp:71]: Fair. Please create a Jira issue - if the pattern is this prevalent, it'll be a nice bit of refactoring to do at some point.

File: apps/constellation/manifest_builder.cpp (hunk 3 above)
Comment (wkaluza) [reply apps/constellation/manifest_builder.cpp:102]: I think we misunderstand each other - I didn't mean testing values of env variables, but injecting envvariables in a unit test context to check that they override defaults and are overridden by cli arguments. Raise the issue - it's a change we can make on top of this after it's merged.

**Review by wkaluza — DISMISSED** [review]

**Review by wkaluza — APPROVED** [review]
